### PR TITLE
Fixed #33767 -- Fixed crash when ordering with F-expression pointing to literal values.

### PR DESCRIPTION
**Pull request by apollo13**

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -330,8 +330,21 @@ def _order_by_pairs(self):
                 if not self.query.standard_ordering:
                     field = field.copy()
                     field.reverse_ordering()
+                src = field.get_source_expressions()[0]
+                if isinstance(src, F):
+                    try:
+                        resolved = src.resolve_expression(self.query, summarize=True)
```
Comment (apollo13) [inline django/db/models/sql/compiler.py:336]: Do we really need to resolve this at this level? Or is there a way to mark it as reference without resolving?

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -330,8 +330,21 @@ def _order_by_pairs(self):
                 if not self.query.standard_ordering:
                     field = field.copy()
                     field.reverse_ordering()
+                src = field.get_source_expressions()[0]
+                if isinstance(src, F):
```
Comment (felixxm) [inline django/db/models/sql/compiler.py:334]: `field` is always an `OrderBy` instance so we can use `.expression`:
```suggestion
                if isinstance(field.expression, F):
```

File: tests/ordering/tests.py [hunk 3]
```diff
@@ -480,6 +480,17 @@ def test_order_by_f_expression_duplicates(self):
             attrgetter("headline"),
         )
 
+    def test_order_by_f_expression_alias(self):
+        qs = (
+            Article.objects.filter(id__lt=0)
+            .annotate(test=Value(42))
+            .order_by(F("test").asc())
+        )
+        self.assertQuerysetEqual(qs, [])
```
Comment (felixxm) [inline tests/ordering/tests.py:489]:
```suggestion
        self.assertSequenceEqual(qs, [])
```

File: tests/ordering/tests.py [hunk 4]
```diff
@@ -480,6 +480,17 @@ def test_order_by_f_expression_duplicates(self):
             attrgetter("headline"),
         )
 
+    def test_order_by_f_expression_alias(self):
+        qs = (
+            Article.objects.filter(id__lt=0)
+            .annotate(test=Value(42))
+            .order_by(F("test").asc())
```
Comment (felixxm) [inline tests/ordering/tests.py:487]: `asc()` is unnecessary:
```suggestion
            .order_by(F("test"))
```

File: tests/ordering/tests.py [hunk 5]
```diff
@@ -480,6 +480,17 @@ def test_order_by_f_expression_duplicates(self):
             attrgetter("headline"),
         )
 
+    def test_order_by_f_expression_alias(self):
```
Comment (felixxm) [inline tests/ordering/tests.py:483]: Maybe:
```suggestion
    def test_order_by_f_expression_to_constant_value(self):
```

**Review by felixxm — COMMENTED** [review]
@apollo13 Thanks :+1: (see related ticket-26192 and [PR](https://github.com/django/django/pull/11396)).

File: django/db/models/sql/compiler.py (hunk 1 above)
Comment (charettes) [reply django/db/models/sql/compiler.py:336]: This seems to be on the right path but I'm not sure `query.annotations` is correct as some might have been excluded through `.values`? Do we want to check `.select` and `.extra_select` as well?

There must be a way to merge this logic with the one below that already cover all of these paths.

**Comment by felixxm** [comment]
@apollo13 Do you have time to keep working on this?

**Comment by apollo13** [comment]
I will pick this up again if needed, but for now I will focus on getting order by / group by aliased.

**Pull request closed without merging**